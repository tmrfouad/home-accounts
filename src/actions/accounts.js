import database from '../firebase/firebase';

// Add_ACCOUNT
export const addAccount = account => ({
  type: 'ADD_ACCOUNT',
  account
});

export const startAddAccount = (accountData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { name = '' } = accountData;

    const account = {
      name
    };

    return database
      .ref(`users/${uid}/accounts`)
      .push(account)
      .then(ref => {
        dispatch(addAccount({ id: ref.key, ...account }));
      })
      .catch(() => {});
  };
};

// REMOVE_ACCOUNT
export const removeAccount = ({ id }) => ({
  type: 'REMOVE_ACCOUNT',
  id
});

export const startRemoveAccount = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/accounts/${id}`)
      .remove()
      .then(() => {
        dispatch(removeAccount({ id }));
      });
  };
};

// EDIT_ACCOUNT
export const editAccount = ({ id } = {}, updates) => ({
  type: 'EDIT_ACCOUNT',
  id,
  updates
});

export const startEditAccount = ({ id } = {}, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/accounts/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editAccount({ id }, updates));
      });
  };
};

// SET_ACCOUNTS
export const setAccounts = accounts => ({
  type: 'SET_ACCOUNTS',
  accounts
});

export const startSetAccounts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/accounts`).once('value', snap => {
      const accounts = [];
      snap.forEach(childSnap => {
        accounts.push({
          id: childSnap.key,
          ...childSnap.val()
        });
      });
      dispatch(setAccounts(accounts));
    });
  };
};
