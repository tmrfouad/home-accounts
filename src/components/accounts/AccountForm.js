import React from 'react';
import { connect } from 'react-redux';
import {
  startAddAccount,
  startEditAccount,
  startRemoveAccount
} from '../../actions/accounts';
import ConfirmModal from '../ConfirmModal';
import { BackButton } from '../BackButton';

export class AccountForm extends React.Component {
  constructor(props) {
    super(props);
    const account = this.props.account;
    this.state = {
      name: account ? account.name : '',
      error: '',
      mode: account ? 'edit' : 'add',
      isModalOpen: false
    };
  }

  onNameChange = e => {
    this.setState({ name: e.target.value });
  };

  onFormSubmit = e => {
    e.preventDefault();
    if (!this.state.name) {
      this.setState({ error: 'Please enter a name!' });
    } else {
      if (this.state.mode === 'edit') {
        this.props.startEditAccount(this.props.account.id, {
          name: this.state.name
        });
      } else {
        if (this.props.accounts.find(acc => acc.name === this.state.name)) {
          this.setState({ error: 'This account already exists!' });
        } else {
          this.props.startAddAccount({
            name: this.state.name
          });
        }
      }
      this.props.history.push('/accounts');
    }
  };

  removeAccount = () => {
    this.props.startRemoveAccount(this.props.account.id);
    this.props.history.push('/accounts');
  };

  openRemoveItemDialog = e => {
    e.preventDefault();
    this.setState(() => ({ isModalOpen: true }));
  };

  closeRemoveItemDialog = () => {
    this.setState(() => ({ isModalOpen: false }));
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div
            className={
              this.props.styles.collapsed
                ? 'content-container content-container--collapsed'
                : 'content-container'
            }
          >
            <BackButton />
            <h2 className="page-header__title">
              {this.state.mode === 'edit' ? 'Edit' : 'Add'} Account
            </h2>
          </div>
        </div>
        <form
          className={
            this.props.styles.collapsed
              ? 'form content-container content-container--collapsed'
              : 'form content-container'
          }
          onSubmit={this.onFormSubmit}
        >
          {this.state.error && (
            <div className="form__error">{this.state.error}</div>
          )}
          <input
            type="text"
            className="text-input"
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <div>
            <button className="button form__action">Save Account</button>
            {this.state.mode === 'edit' && (
              <button
                className="button form__action button--secondary"
                onClick={this.openRemoveItemDialog}
              >
                Remove Account
              </button>
            )}
          </div>
        </form>
        <ConfirmModal
          id="confirmModal"
          messageTitle="Remove Account!"
          messageBody="Are you sure you want to remove this item?"
          isOpen={this.state.isModalOpen}
          onModalClose={this.closeRemoveItemDialog}
          onModalOk={this.removeAccount}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  account: state.accounts.find(t => t.id === props.match.params.id),
  accounts: state.accounts,
  styles: state.styles
});

const mapDispatchToProps = dispatch => ({
  startAddAccount: account => dispatch(startAddAccount(account)),
  startEditAccount: (id, account) =>
    dispatch(startEditAccount({ id }, account)),
  startRemoveAccount: id => dispatch(startRemoveAccount({ id }))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountForm);
