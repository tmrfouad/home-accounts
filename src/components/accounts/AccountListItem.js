import React from 'react';
import { Link } from 'react-router-dom';

const AccountListItem = ({ id, name }) => (
  <Link className="list-item" to={`/accountform/${id}`}>
    <div>
      <h3 className="list-item__title">{name}</h3>
    </div>
  </Link>
);

export default AccountListItem;
