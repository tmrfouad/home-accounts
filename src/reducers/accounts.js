// Accounts Reducer

const accountsReducerDefaultState = [];

export default (state = accountsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_ACCOUNT':
      return [...state, action.account];
    case 'REMOVE_ACCOUNT':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_ACCOUNT':
      return state.map(
        account =>
          account.id === action.id ? { ...account, ...action.updates } : account
      );
    case 'SET_ACCOUNTS':
      return action.accounts;
    default:
      return state;
  }
};
