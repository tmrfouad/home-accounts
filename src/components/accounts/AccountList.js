import React from 'react';
import { connect } from 'react-redux';
import AccountListItem from './AccountListItem';
import { Link } from 'react-router-dom';

export class AccountList extends React.Component {
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Accounts</h2>
            <div className="page-header__actions">
              <Link className="button" to="/accountform">
                Add Account
              </Link>
            </div>
          </div>
        </div>
        <div className="content-container">
          <div className="list-header">
            <div>Name</div>
          </div>
          <div className="list-body">
            {this.props.accounts.length === 0 ? (
              <div className="list-item list-item--message">
                <span>No Accounts</span>
              </div>
            ) : (
              this.props.accounts.map(acc => (
                <AccountListItem key={acc.id} {...acc} />
              ))
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  accounts: state.accounts
});

export default connect(mapStateToProps)(AccountList);
