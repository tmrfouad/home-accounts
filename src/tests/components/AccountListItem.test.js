import React from 'react';
import { shallow } from 'enzyme';
import accounts from '../fixtures/accounts';
import AccountListItem from '../../components/AccountListItem';

test('should render AccountListItem correctly', () => {
  const wrapper = shallow(
    <AccountListItem key={accounts[0].id} {...accounts[0]} />
  );
  expect(wrapper).toMatchSnapshot();
});
