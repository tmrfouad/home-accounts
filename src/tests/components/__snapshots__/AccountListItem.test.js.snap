// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AccountListItem correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/accountform/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Account 1
    </h3>
  </div>
</Link>
`;
